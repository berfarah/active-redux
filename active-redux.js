import _Model from './model';
export { _Model as Model };
import _Registry from './registry';
export { _Registry as Registry };
import * as _Attr from './attributes';
export { _Attr as Attr };

export * from './redux/local';
export * from './redux/remote';
