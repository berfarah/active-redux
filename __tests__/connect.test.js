import * as AR from 'active-redux';
import { Provider } from 'react-redux';
import React from 'react';
import mockStore from 'fixtures/store';
import { mount } from 'enzyme';

const Base = () => <div />;

AR.bind(mockStore);
const Person = AR.define('people', class Person {});

const waitFor = (delay) => new Promise((resolve) => {
  setTimeout(resolve, delay);
});

const findPerson = (delay, error) => () => Promise.all([
  Person.find({ id: '9' }, { remote: false }),
  waitFor(delay),
]).then((data) => (
  error ? Promise.reject(error) : data[0]
));

const wrap = (Subject) => mount(
  <Provider store={mockStore}><Subject /></Provider>
).find(Base);

describe('connect', () => {
  test('loading success', async () => {
    const Subject = AR.connect({
      person: findPerson(200),
    }, {
      delay: 100,
    })(Base);
    const wrapper = wrap(Subject);

    expect(wrapper).toMatchSnapshot(); // Loading
    await waitFor(100);
    expect(wrapper).toMatchSnapshot(); // Past Delay
    await waitFor(100);
    expect(wrapper).toMatchSnapshot(); // Done
  });

  test('loading timeout', async () => {
    const Subject = AR.connect({
      person: findPerson(300),
    }, {
      delay: 100,
      timeout: 200,
    })(Base);
    const wrapper = wrap(Subject);

    expect(wrapper).toMatchSnapshot(); // Loading
    await waitFor(100);
    expect(wrapper).toMatchSnapshot(); // Past Delay
    await waitFor(100);
    expect(wrapper).toMatchSnapshot(); // Timeout
  });

  test('loading error', async () => {
    const Subject = AR.connect({
      person: findPerson(50, new Error('whoops')),
    })(Base);
    const wrapper = wrap(Subject);

    expect(wrapper).toMatchSnapshot(); // Loading
    await waitFor(100);
    expect(wrapper).toMatchSnapshot(); // Error
  });

  test('store changed', async () => {
    const Subject = AR.connect({
      person: findPerson(50),
    })(Base);
    const wrapper = wrap(Subject);

    expect(wrapper).toMatchSnapshot(); // Loading
    await waitFor(100);
    expect(wrapper).toMatchSnapshot(); // Loaded #1
    mockStore.dispatch(AR.api.apiClear('people'));
    await waitFor(100);
    expect(wrapper).toMatchSnapshot(); // Loaded #2
  });
});
