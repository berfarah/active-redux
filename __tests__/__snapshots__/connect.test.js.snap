// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connect loading error 1`] = `
<Base
  dispatch={[Function]}
  errored={Object {}}
  hasDelayed={false}
  hasTimedOut={false}
  isLoading={true}
>
  <div />
</Base>
`;

exports[`connect loading error 2`] = `
<Base
  dispatch={[Function]}
  errored={
    Object {
      "person": true,
    }
  }
  hasDelayed={false}
  hasTimedOut={false}
  isLoading={false}
  person={[Error: whoops]}
>
  <div />
</Base>
`;

exports[`connect loading success 1`] = `
<Base
  dispatch={[Function]}
  errored={Object {}}
  hasDelayed={false}
  hasTimedOut={false}
  isLoading={true}
>
  <div />
</Base>
`;

exports[`connect loading success 2`] = `
<Base
  dispatch={[Function]}
  errored={Object {}}
  hasDelayed={true}
  hasTimedOut={false}
  isLoading={true}
>
  <div />
</Base>
`;

exports[`connect loading success 3`] = `
<Base
  dispatch={[Function]}
  errored={
    Object {
      "person": false,
    }
  }
  hasDelayed={true}
  hasTimedOut={false}
  isLoading={false}
  person={
    Person {
      "data": Object {
        "attributes": Object {
          "address": Object {
            "road": "123 Main St",
          },
          "first-name": "Dan",
          "last-name": "Gebhardt",
          "twitter": "dgeb",
        },
        "id": "9",
        "links": Object {
          "self": "http://example.com/people/9",
        },
        "relationships": Object {
          "comments": Object {
            "data": Array [
              Object {
                "id": "5",
                "type": "comments",
              },
              Object {
                "id": "12",
                "type": "comments",
              },
            ],
          },
        },
        "type": "people",
      },
    }
  }
>
  <div />
</Base>
`;

exports[`connect loading timeout 1`] = `
<Base
  dispatch={[Function]}
  errored={Object {}}
  hasDelayed={false}
  hasTimedOut={false}
  isLoading={true}
>
  <div />
</Base>
`;

exports[`connect loading timeout 2`] = `
<Base
  dispatch={[Function]}
  errored={Object {}}
  hasDelayed={true}
  hasTimedOut={false}
  isLoading={true}
>
  <div />
</Base>
`;

exports[`connect loading timeout 3`] = `
<Base
  dispatch={[Function]}
  errored={Object {}}
  hasDelayed={true}
  hasTimedOut={true}
  isLoading={true}
>
  <div />
</Base>
`;

exports[`connect store changed 1`] = `
<Base
  dispatch={[Function]}
  errored={Object {}}
  hasDelayed={false}
  hasTimedOut={false}
  isLoading={true}
>
  <div />
</Base>
`;

exports[`connect store changed 2`] = `
<Base
  dispatch={[Function]}
  errored={
    Object {
      "person": false,
    }
  }
  hasDelayed={false}
  hasTimedOut={false}
  isLoading={false}
  person={
    Person {
      "data": Object {
        "attributes": Object {
          "address": Object {
            "road": "123 Main St",
          },
          "first-name": "Dan",
          "last-name": "Gebhardt",
          "twitter": "dgeb",
        },
        "id": "9",
        "links": Object {
          "self": "http://example.com/people/9",
        },
        "relationships": Object {
          "comments": Object {
            "data": Array [
              Object {
                "id": "5",
                "type": "comments",
              },
              Object {
                "id": "12",
                "type": "comments",
              },
            ],
          },
        },
        "type": "people",
      },
    }
  }
>
  <div />
</Base>
`;

exports[`connect store changed 3`] = `
<Base
  dispatch={[Function]}
  errored={
    Object {
      "person": false,
    }
  }
  hasDelayed={true}
  hasTimedOut={false}
  isLoading={false}
>
  <div />
</Base>
`;
